### PR TITLE
Generalise the view-change test

### PR DESCRIPTION
**Pull request by kc1212**
This is to prepare for the upcoming work on improving viewchange to work
in more failures scenarios.

### CODE REVIEW

File: omniledger/service/service_test.go [hunk 1]
```diff
@@ -756,63 +756,77 @@ func TestService_SetBadConfig(t *testing.T) {
...
-func TestService_RotateLeader(t *testing.T) {
+// TestService_ViewChange is an end-to-end test for view-change. We kill the
+// first nFailures nodes, where the nodes at index 0 is the current leader. The
+// node at index nFailures should become the new leader. Then, we try to send a
+// transaction to a follower, at index nFailures+1. The new leader (at index
+// nFailures) should poll for new transactions and eventually make a new block
+// containing that transaction. The new transaction should be stored on all
+// followers. Finally, we bring the failed nodes back up and they should
+// contain the transactions that they missed (TODO this does not work yet
+// because we the config contract fails).
```
Comment (jeffallen) [inline omniledger/service/service_test.go]: s/we//

File: omniledger/service/service_test.go [hunk 2]
```diff
@@ -756,63 +756,77 @@ func TestService_SetBadConfig(t *testing.T) {
...
 	defer s.local.CloseAll()
 
 	for _, service := range s.services {
 		service.SetPropagationTimeout(interval / 2)
 	}
 
-	// Stop the leader, then the next node should take over.
-	s.service().TestClose()
-	s.hosts[0].Pause()
+	// The first nFailures hosts then the next node at index nFailures
```
Comment (jeffallen) [inline omniledger/service/service_test.go]: // Stop the first nFailures hosts.

File: omniledger/service/service_test.go [hunk 3]
```diff
@@ -756,63 +756,77 @@ func TestService_SetBadConfig(t *testing.T) {
...
+	// should take over.
+	for i := 0; i < nFailures; i++ {
+		s.services[i].TestClose()
+		s.hosts[i].Pause()
+	}
 
 	// wait for the new block
```
Comment (jeffallen) [inline omniledger/service/service_test.go]: // Wait for proof that the new expected leader, s.services[nFailures], has taken over.

**Comment by kc1212** [comment]
Thanks, fixed.

File: omniledger/service/service_test.go [hunk 4]
```diff
@@ -756,63 +756,77 @@ func TestService_SetBadConfig(t *testing.T) {
...
+	// should take over.
+	for i := 0; i < nFailures; i++ {
+		s.services[i].TestClose()
+		s.hosts[i].Pause()
+	}
 
-	// wait for the new block
+	// Wait for proof that the new expected leader, s.services[nFailures], has taken over.
 	var ok bool
 	for i := 0; i < 5; i++ {
```
Comment (ineiti) [inline omniledger/service/service_test.go]: Shouldn't this just

```
time.Sleep(2 * nFailures * s.interval)
```

and then verify right away if the new leader is active?

Looping here seems too fuzzy to me.

Comment (kc1212) [reply omniledger/service/service_test.go]: nice catch! and you're right, we don't need looping.

**Pull request merged**